### PR TITLE
NNCLR Implementation

### LINKED ISSUE #316: Implement NNCLR
# 🌟 New SSL approach addition

## Approach description

NNCLR: https://arxiv.org/abs/2104.14548

> Self-supervised learning algorithms based on instance discrimination train encoders to be invariant to pre-defined transformations of the same instance. While most methods treat different views of the same image as positives for a contrastive loss, we are interested in using positives from other instances in the dataset. Our method, Nearest-Neighbor Contrastive Learning of visual Representations (NNCLR), samples the nearest neighbors from the dataset in the latent space, and treats them as positives. This provides more semantic variations than pre-defined transformations.
We find that using the nearest-neighbor as positive in contrastive losses improves performance significantly on ImageNet classification, from 71.7% to 75.6%, outperforming previous state-of-the-art methods. On semi-supervised learning benchmarks we improve performance significantly when only 1% ImageNet labels are available, from 53.8% to 56.5%. On transfer learning benchmarks our method outperforms state-of-the-art methods (including supervised learning with ImageNet) on 8 out of 12 downstream datasets. Furthermore, we demonstrate empirically that our method is less reliant on complex data augmentations. We see a relative reduction of only 2.1% ImageNet Top-1 accuracy when we train using only random crops.

## Pseudocode

```python
# f: backbone encoder + projection MLP
# g: prediction MLP
# Q: queue

for x in loader: # load a minibatch x with n samples
    x1, x2 = aug(x), aug(x) # random augmentation
    z1, z2 = f(x1), f(x2) # projections, n-by-d
    p1, p2 = g(z1), g(z2) # predictions, n-by-d

    NN1 = NN(z1, Q) # top-1 NN lookup, n-by-d
    NN2 = NN(z2, Q) # top-1 NN lookup, n-by-d

    loss = L(NN1, p2)/2 + L(NN2, p1)/2
    loss.backward() # back-propagate
    update(f, g) # SGD update
    update_queue(Q, z1) # Update queue with latest projection embeddings

def L(nn, p, temperature=0.1):
    nn = normalize(nn, dim=1) # l2-normalize
    p = normalize(p, dim=1) # l2-normalize

    logits = nn @ p.T # Matrix multiplication, n-by-n
    logits /= temperature # Scale by temperature

    n = p.shape[0] # mini-batch size
    labels = range(n)

    loss = cross_entropy(logits, labels)

    return loss
```
## Open source status

* [ ] the model implementation is available: (give details)
* [ ] the model weights are available: (give details)
* [ ] Debidatta Dwibedi, Yusuf Aytar, Jonathan Tompson, Pierre Sermanet, Andrew Zisserman

### PR DESCRIPTION
**Pull request by OlivierDehaene**
## Required

- [X] NNCLRLoss and Criterion
- [ ] Documentation
  - [ ] Loss
  - [ ] SSL Approaches + Index
  - [ ] Model Zoo
  - [ ] Project
- [ ] Default configs
    - [ ] pretrain
    - [X] test/integration
    - [X] debugging/pretrain
- [ ] Benchmarks
  - [ ] ImageNet
  - [ ] Imagenette

closes #316

### CODE REVIEW

File: vissl/hooks/nnclr_hooks.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+import logging
+
+import torch
+from classy_vision import tasks
+from classy_vision.hooks.classy_hook import ClassyHook
+from vissl.models.heads import MLP
+from vissl.utils.env import get_machine_local_and_dist_rank
+
+
+class NNCLRHook(ClassyHook):
```
Comment (OlivierDehaene) [inline vissl/hooks/nnclr_hooks.py:12]: @prigoyal,

I'm not sure if this is the best way of creating a predictor head.

**Comment by prigoyal** [comment]
Hi @OlivierDehaene , hope things are well :) please lmk if there are any blockers etc that I can help with for this PR. :) I am looking forward to having this merged in VISSL :)

**Comment by OlivierDehaene** [comment]
@prigoyal,

Sorry, I did not found any time to work on this recently. I'm still figuring out if there is a problem in this implementation as it leads to subpar performance on Imagenette 180.

**Comment by prigoyal** [comment]
Hi @OlivierDehaene , just wanted to check in: what's your plan for this PR ? Is there anything that we can help with? :)

**Comment by OlivierDehaene** [comment]
Hello @prigoyal, I'm running a few tests but it should be close to completion now.

File: vissl/losses/nnclr_info_nce_loss.py [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+        )
+
+        logits_ab = (
+            torch.matmul(nn_a, gather_from_all(predictions_b).T) / self.temperature
+        )
+        logits_ba = (
+            torch.matmul(nn_b, gather_from_all(predictions_a).T) / self.temperature
+        )
+
+        loss = (
```
Comment (OlivierDehaene) [inline vissl/losses/nnclr_info_nce_loss.py:201]: @prigoyal,
For now I only use half of the negatives as discussed here: https://github.com/facebookresearch/vissl/issues/316#issuecomment-836708548
I'm running a few tests to see how much it affects performance.

**Comment by clane9** [comment]
Hi @OlivierDehaene and @prigoyal , hope you're well :). I'm interested in seeing NNCLR make it into VISSL. Let me know if you could use help on this PR, I'd be happy to contribute! And thanks for making this work available!

**Comment by OlivierDehaene** [comment]
Hello @prigoyal, @clane9,
Sorry for ghosting, I took a new position in September and did not find the time to work on personal/OSS projects. Maybe this PR should be closed and the task assigned to a new person?

**Comment by OlivierDehaene** [comment]
Hello @prigoyal, @clane9,
Sorry for not answering, I took a new position in September and did not find the time to work on personal/OSS projects. Maybe this PR should be closed and the task assigned to a new person?
